### PR TITLE
chore: remove decimals value from futures objects

### LINKED ISSUE #6645: data source no. of decimals to be stored in the data source definition
The number of decimals on a data source is not currently stored in the data source definition ([as described here](https://github.com/vegaprotocol/specs/blob/master/protocol/0045-DSRC-data_sourcing.md#allowable-types)) it is stored in the  market definition.

To align with the specification we need to store the number of decimals on a data source in the data source definition.

### PR DESCRIPTION
**Pull request by AnExsomnis**
close: #6645

### CODE REVIEW

File: protos/sources/vega/data/v1/data.proto [hunk 1]
```diff
@@ -31,6 +31,9 @@ message Property {
   string name = 1;
   // value is the value of the property.
   string value = 2;
+  // An optional decimal place to be be applied on the provided value
+  // valid only for PropertyType of Type DECIMAL and INTEGER
```
Comment (candida-d) [inline protos/sources/vega/data/v1/data.proto:35]:
```suggestion
  // An optional decimal place to be applied on the provided value
  // valid only for PropertyType of type DECIMAL and INTEGER
```

File: protos/sources/vega/data/v1/spec.proto [hunk 2]
```diff
@@ -19,6 +19,10 @@ message PropertyKey {
   string name = 1;
   // type is the type of the property.
   Type type = 2;
+  // An optional decimal place to be be applied on the provided value
+  // valid only for PropertyType of Type DECIMAL and INTEGER
```
Comment (candida-d) [inline protos/sources/vega/data/v1/spec.proto:23]:
```suggestion
  // An optional decimal place to be applied on the provided value
  // valid only for PropertyType of type DECIMAL and INTEGER
```

Comment (candida-d) [reply protos/sources/vega/data/v1/spec.proto:23]: Just wondering out of curiosity - why is there a spec.proto and a data.proto with the same types?

Comment (AnExsomnis) [reply protos/sources/vega/data/v1/spec.proto]: Both corrections are added, thanks!
The types are created for different purpose - one is used in the core and the other one in the datanode part.
Usually for the datanode part are used types that correspond with existing ones in the core - but simplified.
Because the core ones need to handle the core respective logic, and the ones in the datanode - do not.

File: CHANGELOG.md [hunk 3]
```diff
@@ -37,6 +37,7 @@
 - [6819](https://github.com/vegaprotocol/vega/issues/6819) - Support long-living token in wallet client API
 - [6964](https://github.com/vegaprotocol/vega/issues/6964) - Add support for long living tokens with expiry
 - [6991](https://github.com/vegaprotocol/vega/issues/6991) - Expose error field in explorer API
+- [6645](https://github.com/vegaprotocol/vega/issues/6645) - Set decimal number value to be used from oracle instead of from tradable instruments
```
Comment (ze97286) [inline CHANGELOG.md:40]: I think it's a breaking change isn't it?

File: protos/sources/vega/data/v1/data.proto [hunk 4]
```diff
@@ -31,6 +31,9 @@ message Property {
   string name = 1;
   // value is the value of the property.
   string value = 2;
+  // An optional decimal place to be be applied on the provided value
+  // valid only for PropertyType of type DECIMAL and INTEGER
+  uint32 number_decimal_places = 3;
```
Comment (ze97286) [inline protos/sources/vega/data/v1/data.proto:36]: why is this necessary? we already know it from the spec

File: protos/sources/vega/data/v1/spec.proto [hunk 5]
```diff
@@ -19,6 +19,10 @@ message PropertyKey {
   string name = 1;
   // type is the type of the property.
   Type type = 2;
+  // An optional decimal place to be be applied on the provided value
+  // valid only for PropertyType of type DECIMAL and INTEGER
```
Comment (ze97286) [inline protos/sources/vega/data/v1/spec.proto:23]: is it actually relevant for decimal?

File: CHANGELOG.md (hunk 3 above)
Comment (AnExsomnis) [reply CHANGELOG.md]: Yes, sorry I missed that.
Thanks!

File: protos/sources/vega/data/v1/spec.proto (hunk 5 above)
Comment (ze97286) [reply protos/sources/vega/data/v1/spec.proto:23]: I think it is but I don't think it's handled correctly (or at all), I'll comment about it somewhere else.

File: protos/sources/vega/data/v1/spec.proto [hunk 6]
```diff
@@ -19,6 +19,10 @@ message PropertyKey {
...
+  uint32 number_decimal_places = 3;
```
Comment (ze97286) [inline protos/sources/vega/data/v1/spec.proto:24]: should this be optional

Comment (AnExsomnis) [reply protos/sources/vega/data/v1/spec.proto:24]: I haven't set this as optional, because we take this value only if the type is integer or decimal.
If it is not set - will be 0 when deserialized.
Do you think its better to be set as optional here?

File: protos/sources/vega/data/v1/data.proto (hunk 4 above)
Comment (AnExsomnis) [reply protos/sources/vega/data/v1/data.proto:36]: Yes, that decimal will be already applied in the string value.

File: protos/sources/vega/data/v1/spec.proto (hunk 6 above)
Comment (jeremyletang) [reply protos/sources/vega/data/v1/spec.proto:24]: optional would make sense, specifically because it is not mandatory to have it as you say?

Comment (AnExsomnis) [reply protos/sources/vega/data/v1/spec.proto:24]: all right

File: core/oracles/oracle_spec.go [hunk 7]
```diff
@@ -40,6 +40,8 @@ var (
 	ErrMissingPropertyName = errors.New("a property name is required")
 	// ErrInvalidPropertyKey is returned if validation finds a reserved Vega property key.
 	ErrInvalidPropertyKey = errors.New("property key is reserved")
+	// ErrMultipleUniqueNamesInFilterList is returned if filters with same key names exists inside a single list.
+	ErrMultipleUniqueNamesInFilterList = errors.New("multiple unique keys with same name found in filter list")
```
Comment (ze97286) [inline core/oracles/oracle_spec.go:44]: maybe remove the *unique*

Comment (AnExsomnis) [reply core/oracles/oracle_spec.go]: Made it like this everywhere:
`
ErrMultipleSameKeyNamesInFilterList = errors.New("multiple keys with same name found in filter list")

`

File: core/types/data_source.go [hunk 8]
```diff
@@ -2,13 +2,16 @@ package types
...
 	"encoding/hex"
+	"errors"
 	"fmt"
 
 	"code.vegaprotocol.io/vega/libs/crypto"
 	"code.vegaprotocol.io/vega/libs/proto"
 	vegapb "code.vegaprotocol.io/vega/protos/vega"
 )
 
+var ErrMultipleUniqueNamesInFilterList = errors.New("multiple unique keys with same name found in filter list")
```
[the remark this answers is not in the document]
Comment (AnExsomnis) [reply core/types/data_source.go]: This has been addressed, please see https://github.com/vegaprotocol/vega/pull/6935#discussion_r1037575301

File: protos/sources/vega/data/v1/spec.proto (hunk 6 above)
Comment (AnExsomnis) [reply protos/sources/vega/data/v1/spec.proto]: Set the fields to optional everywhere.

File: protos/sources/vega/data/v1/data.proto (hunk 4 above)
Comment (AnExsomnis) [reply protos/sources/vega/data/v1/data.proto]: This message is deserialized into this object from the datanode side - https://github.com/vegaprotocol/vega/blob/develop/datanode/entities/data.go#L32

File: core/products/future.go [hunk 9]
```diff
@@ -243,13 +243,27 @@ func NewFuture(ctx context.Context, log *logging.Logger, f *types.Future, oe Ora
...
 	}
 
+	decimals := uint32(0)
+	dSrcSpec := f.DataSourceSpecForSettlementData.Data.GetDataSourceSpecConfiguration()
+	if dSrcSpec != nil {
+		for _, f := range dSrcSpec.Filters {
+			// Oracle specs with more than one unique filter names are not allowed to exists, so we do not have to make that check here.
+			// We are good to only check if the type is `PropertyKey_TYPE_DECIMAL` or `PropertyKey_TYPE_INTEGER`, because we take decimals
+			// into consideration only in those cases.
+			if f.Key.Type == datapb.PropertyKey_TYPE_DECIMAL || f.Key.Type == datapb.PropertyKey_TYPE_INTEGER {
```
[the remark this answers is not in the document]
Comment (AnExsomnis) [reply core/products/future.go]: Yes, the settlement data is always supposed to be a numeric value. 
And inside the the callback `updateSettlementData` is updated the value as uint, you mean not checking if it is decimal otr integer?

[the remark this answers is not in the document]
Comment (AnExsomnis) [reply core/products/future.go]: Yes, this part of the code exists form before and since we are not handling decimals at the moment, that is why its not touched.
But I'll fix it.

[the remark this answers is not in the document]
Comment (AnExsomnis) [reply core/products/future.go]: The change for handling the two types inside the updating settlement data has been added.

File: core/governance/market.go [hunk 10]
```diff
@@ -271,10 +270,23 @@ func validateFuture(future *types.FutureProduct, decimals uint64, assets Assets,
...
-		return types.ProposalErrorInvalidFutureProduct, fmt.Errorf("invalid oracle spec binding for settlement data: %w", err)
+
+	switch future.DataSourceSpecBinding.SettlementDataProperty {
+	case datapb.PropertyKey_TYPE_DECIMAL.String():
+		err := ospec.EnsureBoundableProperty(future.DataSourceSpecBinding.SettlementDataProperty, datapb.PropertyKey_TYPE_DECIMAL)
+		if err != nil {
+			return types.ProposalErrorInvalidFutureProduct, fmt.Errorf("invalid oracle spec binding for settlement data: %w", err)
+		}
+	case datapb.PropertyKey_TYPE_INTEGER.String():
+		fallthrough
```
Comment (EVODelavega) [inline core/governance/market.go:281]: Why specify the case if it's a fallthrough to the default?

File: core/governance/market.go [hunk 11]
```diff
@@ -453,15 +465,27 @@ func validateUpdateFuture(future *types.UpdateFutureProduct, et *enactmentTime)
...
```
Comment (EVODelavega) [inline core/governance/market.go:480]: same

File: core/products/future.go [hunk 12]
```diff
@@ -209,24 +203,46 @@ func (f *Future) updateSettlementData(ctx context.Context, data oracles.OracleDa
...
+				"could not parse decimal type property acting as settlement data",
+				logging.Error(err),
+			)
+			return err
+		}
+
+		odata.settlData.AsDecimal = &settlDataAsDecimal
+
+	case datapb.PropertyKey_TYPE_INTEGER:
+		fallthrough
```
Comment (EVODelavega) [inline core/products/future.go:223]: same here

File: core/products/future.go [hunk 13]
```diff
@@ -243,13 +259,28 @@ func NewFuture(ctx context.Context, log *logging.Logger, f *types.Future, oe Ora
 		return nil, err
 	}
 
+	dSrcSpec := f.DataSourceSpecForSettlementData.Data.GetDataSourceSpecConfiguration()
+	if dSrcSpec != nil {
+		for _, f := range dSrcSpec.Filters {
+			// Oracle specs with more than one unique filter names are not allowed to exists, so we do not have to make that check here.
+			// We are good to only check if the type is `PropertyKey_TYPE_DECIMAL` or `PropertyKey_TYPE_INTEGER`, because we take decimals
+			// into consideration only in those cases.
+			if f.Key.Type == datapb.PropertyKey_TYPE_DECIMAL || f.Key.Type == datapb.PropertyKey_TYPE_INTEGER && f.Key.NumberDecimalPlaces != nil {
```
Comment (EVODelavega) [inline core/products/future.go:268]: This looks off to me, shouldn't we have brackets around the integer + decimal places part?

File: core/products/future_for_test.go [hunk 14]
```diff
@@ -16,11 +16,14 @@ import (
 	"context"
 	"strconv"
 
+	datapb "code.vegaprotocol.io/vega/protos/vega/data/v1"
+
 	"code.vegaprotocol.io/vega/core/oracles"
 )
 
-func (f *Future) SetSettlementData(ctx context.Context, priceName string, settlementData uint64) {
+func (f *Future) SetSettlementData(ctx context.Context, priceName string, settlementData uint64, settlementDataType datapb.PropertyKey_Type) {
```
Comment (EVODelavega) [inline core/products/future_for_test.go:24]: Should this function still take a `uint64`? If settlement data is a decimal, then this doesn't really make much sense

File: core/products/future_test.go [hunk 15]
```diff
@@ -36,47 +36,125 @@ func TestScalingOfSettlementData(t *testing.T) {
...
+			AsUint: num.NewUint(100000),
+		}, 3,
+	)
 	require.NoError(t, err)
 	require.Equal(t, num.NewUint(1000), scaled)
+
+	// Create test future with settlement data type decimal with decimals (that represents a decimal)
+	ft = testFuture(t, datapb.PropertyKey_TYPE_DECIMAL)
+
+	// settlement data is in 5 decimal places, asset in 3 decimal places => x10^-2
```
Comment (EVODelavega) [inline core/products/future_test.go:108]: This test makes no sense. If the settlement data is already in decimal format, then we should not have a specific number of decimal places. We should take whatever decimal value from the external source, and scale it to the uint in asset precision. In this example, with a value of `10000.00001` (5 decimal places) should yield `1000000001` (if the asset has 5 decimal places), or `100000000100000` (for an asset with 10 decimal places) and so on

File: core/types/data.go [hunk 16]
```diff
@@ -238,36 +239,43 @@ func (s DataSourceSpecSigners) String() string {
 }
 
 type DataSourceSpecPropertyKey struct {
-	Name string
-	Type DataSourceSpecPropertyKeyType
+	Name                string
+	Type                DataSourceSpecPropertyKeyType
+	NumberDecimalPlaces *uint64
```
Comment (EVODelavega) [inline core/types/data.go:244]: Not sure if this is where we should keep this field...

File: core/types/data.go [hunk 17]
```diff
@@ -238,36 +239,43 @@ func (s DataSourceSpecSigners) String() string {
...
+	return pk
 }
 
 func (k *DataSourceSpecPropertyKey) DeepClone() *DataSourceSpecPropertyKey {
 	return &DataSourceSpecPropertyKey{
-		Name: k.Name,
-		Type: k.Type,
+		Name:                k.Name,
+		Type:                k.Type,
+		NumberDecimalPlaces: k.NumberDecimalPlaces,
```
Comment (EVODelavega) [inline core/types/data.go:270]: Even if we should keep this field, this is not a deep clone. the clone would share this pointer with the original object

File: core/products/future.go (hunk 13 above)
Comment (AnExsomnis) [reply core/products/future.go:268]: Not needed I think, because the && is with higher prescendence than ||

File: libs/num/numeric.go [hunk 18]
```diff
@@ -0,0 +1,83 @@
+package num
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+)
+
+type Numeric struct {
+	AsUint    *Uint
```
Comment (EVODelavega) [inline libs/num/numeric.go:10]: I would not export these fields. This is prone to abuse and concurrency issues/bugs due to shared pointers

File: libs/num/numeric.go [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+	AsDecimal *Decimal
+}
+
+func (n *Numeric) Clone() *Numeric {
+	nn := &Numeric{}
+
+	if n.AsUint != nil {
+		nn.AsUint = n.AsUint
```
Comment (EVODelavega) [inline libs/num/numeric.go:18]: The uint should be cloned

File: libs/num/numeric.go [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+		return nn
+	}
+
+	if n.AsDecimal != nil {
+		nn.AsDecimal = n.AsDecimal
```
Comment (EVODelavega) [inline libs/num/numeric.go:23]: create a pointer to a copy

File: libs/num/numeric.go [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+	}
+	if n.AsDecimal != nil {
+		return n.AsDecimal.String()
+	}
+
+	return ""
+}
+
+func NumericFromString(s string) (*Numeric, error) {
+	// Check if the provided string contains a ".", because if it does not,
```
Comment (EVODelavega) [inline libs/num/numeric.go:41]: Start by checking for an empty string

File: libs/num/uint.go [hunk 22]
```diff
@@ -149,6 +149,10 @@ func (u Uint) Uint64() uint64 {
 	return u.u.Uint64()
 }
 
+func (u Uint) Uint32() uint32 {
```
Comment (EVODelavega) [inline libs/num/uint.go:152]: Might be useful to add something like:

```go
func (u Uint) ToInt[T int | int8 | int32 | ...] () T {
     return T(u.Uint64())
}
```
I believe the `constraints` package has a constraint that includes all integers

File: protos/sources/vega/data/v1/spec.proto (hunk 5 above)
Comment (EVODelavega) [reply protos/sources/vega/data/v1/spec.proto:23]: It shouldn't be. The decimal places is relevant for integer values (e.g. settlement price for ETH is expresset in gwei -> decimal places should be specified so we can scale the oracle value to wei correctly). If the data is provided in decimals (ie settlement asset in USD, 1.1 or 1.23) just have to be scaled by multiplying the decimal by 10^asset precision. Combining number of decimal places on a decimal value makes very little sense. The only thing this could possibly do is truncate decimals from the source, which doesn't sit well with me

File: core/governance/market.go (hunk 10 above)
Comment (AnExsomnis) [reply core/governance/market.go]: I left these to aim more clarity/readability when reading the code, but I guess no one needed that (except me); Addressed this and the below 2 as well: https://github.com/vegaprotocol/vega/pull/6935/commits/14c69f3c1ab4f4a6fdbaffe0ee6d971d6658e4f8#diff-1527092608a66a12e24c11a95fc2ca1e9d2b73063b35c78ad06651b45b37bef8R281
Thanks!

File: libs/num/numeric.go (hunk 18 above)
Comment (AnExsomnis) [reply libs/num/numeric.go]: Addressed.
Changed everywhere to set and get via methods.
Thanks!

File: libs/num/numeric.go (hunk 21 above)
Comment (AnExsomnis) [reply libs/num/numeric.go:46]: This has been corrected. Thanks!

File: core/products/future_for_test.go (hunk 14 above)
Comment (AnExsomnis) [reply core/products/future_for_test.go]: Yes, I missed this, has been changed now
Thanks!

File: core/types/data.go (hunk 17 above)
Comment (AnExsomnis) [reply core/types/data.go]: Addressed - the two values has the same address when its done like this -> https://github.com/vegaprotocol/vega/pull/6935/commits/db7f4cd7157e7edeeb1c9c464b9094521d5be6fb#diff-38be7115fe6ae7d810706c7f1c7224e82ecc7da720c0e330d82f869b1e7a7189L261-L265
Thanks!

File: libs/num/numeric.go (hunk 19 above)
Comment (AnExsomnis) [reply libs/num/numeric.go]: Changed this to use the same address for the Uint and a pointer to a copy for decimal -> https://github.com/vegaprotocol/vega/pull/6935/commits/db7f4cd7157e7edeeb1c9c464b9094521d5be6fb#diff-01e84bbc9c49dbdd4b9e45ec724423ea40a7c808577eebb58ed4f69f25ac4755R14-R27
Thanks!
* There are a lot of places in the code where this is not done in the deep clones, should we track and fix them?

File: core/products/future_test.go (hunk 15 above)
Comment (AnExsomnis) [reply core/products/future_test.go:106]: Not sure what do you mean.
The decimal places that are sent via the parameter here are the asset decimals.
When the settlement data type is decimal - the place set inside `NumberOfDecimalPlaces` field are not taken into the calculation -> https://github.com/vegaprotocol/vega/pull/6935/commits/db7f4cd7157e7edeeb1c9c464b9094521d5be6fb#diff-01e84bbc9c49dbdd4b9e45ec724423ea40a7c808577eebb58ed4f69f25ac4755R68-R75
But only the target decimals - in this case 3.

And when the settlement data is updated with a decimal type - they are also not set in the new object -> https://github.com/vegaprotocol/vega/pull/6935/files#diff-72277c761cbd8dabe6ed10e8f5128682486134c673a42959dca45dcff971df3fR221
and upon creation are also not taken -> https://github.com/vegaprotocol/vega/pull/6935/files#diff-72277c761cbd8dabe6ed10e8f5128682486134c673a42959dca45dcff971df3fR269-R271
Or you meant that this test is artificial? But I dont know why it would be

File: libs/num/uint.go (hunk 22 above)
Comment (AnExsomnis) [reply libs/num/uint.go]: Havent done that. As a matter of fact removed this method because it is not used and needed. must have been left from some previous iteration.

File: core/execution/market_snapshot.go [hunk 23]
```diff
@@ -155,7 +155,7 @@ func NewMarketFromSnapshot(
 		marketActivityTracker:      marketActivityTracker,
 		positionFactor:             positionFactor,
 		stateVarEngine:             stateVarEngine,
-		settlementDataInMarket:     em.SettlementData,
+		settlementDataInMarket:     &em.SettlementData,
```
Comment (EVODelavega) [inline core/execution/market_snapshot.go:158]: This doesn't quite make sense. If a field can be a nil pointer in the market, but not on the `ExecMarket`/snapshot type, then restoring a snapshot arguably doesn't restore the _exact_ same state. I believe the snapshot hash would still match in this case, but it doesn't look entirely clean IMO

File: core/execution/market_snapshot.go [hunk 24]
```diff
@@ -177,9 +177,12 @@ func NewMarketFromSnapshot(
 
 func (m *Market) getState() *types.ExecMarket {
 	rf := m.risk.GetRiskFactors()
-	var sp *num.Uint
+	var sp num.Numeric
 	if m.settlementDataInMarket != nil {
-		sp = m.settlementDataInMarket.Clone()
+		spp := m.settlementDataInMarket.Clone()
+		if spp != nil {
```
Comment (EVODelavega) [inline core/execution/market_snapshot.go:183]: How could `Clone` on `settlementDataInMarket` (which we've already confirmed to not be nil at this point) return nil?

File: libs/num/numeric.go [hunk 25]
```diff
@@ -0,0 +1,108 @@
...
+
+	return n
+}
+
+func (n *Numeric) Decimal() *Decimal {
+	return n.asDecimal
+}
+
+func (n *Numeric) Uint() *Uint {
+	return n.asUint
```
Comment (EVODelavega) [inline libs/num/numeric.go:107]: We should make sure to return a clone of the value to avoid the internal values to be updated elsewhere

File: libs/num/numeric.go [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+
+func (n *Numeric) SetDecimal(d *Decimal) *Numeric {
+	n.asDecimal = d
+	n.asUint = nil
+
+	return n
+}
+
+func (n *Numeric) Decimal() *Decimal {
+	return n.asDecimal
```
Comment (EVODelavega) [inline libs/num/numeric.go:103]: Same here: avoid sharing pointers

File: core/products/future.go [hunk 27]
```diff
@@ -65,65 +65,59 @@ type oracle struct {
...
-	settlementData    *num.Uint
+	settlData         *num.Numeric
 	tradingTerminated bool
 }
 
-func (d *oracleData) SettlementData() (*num.Uint, error) {
-	if d.settlementData == nil {
+// SettlementData returns oracle data settlement data scaled as Uint.
+func (o *oracleData) SettlementData(op, ap uint32) (*num.Uint, error) {
+	if o.settlData.Decimal() == nil && o.settlData.Uint() == nil {
```
Comment (EVODelavega) [inline core/products/future.go:77]: Given that these functions should return a copy/clone, it's probably slightly more efficient to have `IsUint` and `IsDecimal` methods here, which just `return n.uint != nil` and `return n.dec != nil`, so we can check the wrapped numeric type without needing to clone the value

File: libs/num/numeric.go (hunk 26 above)
Comment (AnExsomnis) [reply libs/num/numeric.go:103]: I thought you said it should be sharing pointers - like here https://github.com/vegaprotocol/vega/pull/6935#discussion_r1048941709

File: core/products/future.go [hunk 28]
```diff
@@ -209,24 +203,45 @@ func (f *Future) updateSettlementData(ctx context.Context, data oracles.OracleDa
...
+		if err != nil {
+			f.log.Error(
+				"could not parse decimal type property acting as settlement data",
+				logging.Error(err),
+			)
+			return err
+		}
+
+		odata.settlData.SetDecimal(&settlDataAsDecimal)
+		f.oracle.binding.settlementDataDecimals = uint64(0)
```
Comment (EVODelavega) [inline core/products/future.go:221]: ?? why are we setting this to 0 here? If we're dealing with a decimal, at this point we should've already made sure this field is valid (ie set to zer)

File: core/products/future.go [hunk 29]
```diff
@@ -243,13 +258,28 @@ func NewFuture(ctx context.Context, log *logging.Logger, f *types.Future, oe Ora
...
 
+	dSrcSpec := f.DataSourceSpecForSettlementData.Data.GetDataSourceSpecConfiguration()
+	if dSrcSpec != nil {
+		for _, f := range dSrcSpec.Filters {
+			// Oracle specs with more than one unique filter names are not allowed to exists, so we do not have to make that check here.
+			// We are good to only check if the type is `PropertyKey_TYPE_DECIMAL` or `PropertyKey_TYPE_INTEGER`, because we take decimals
+			// into consideration only in those cases.
+			if f.Key.Type == datapb.PropertyKey_TYPE_INTEGER && f.Key.NumberDecimalPlaces != nil {
+				oracleBinding.settlementDataPropertyType = f.Key.Type
+				if f.Key.Type == datapb.PropertyKey_TYPE_INTEGER {
```
Comment (EVODelavega) [inline core/products/future.go:269]: We've already checked the key type in the outer `if`, this can be removed here

File: core/products/future_for_test.go [hunk 30]
```diff
@@ -14,13 +14,20 @@ package products
...
-	"strconv"
+
+	"code.vegaprotocol.io/vega/libs/num"
+	datapb "code.vegaprotocol.io/vega/protos/vega/data/v1"
 
 	"code.vegaprotocol.io/vega/core/oracles"
 )
 
-func (f *Future) SetSettlementData(ctx context.Context, priceName string, settlementData uint64) {
+func (f *Future) SetSettlementData(ctx context.Context, priceName string, settlementData *num.Numeric, settlementDataType datapb.PropertyKey_Type) {
```
Comment (EVODelavega) [inline core/products/future_for_test.go:24]: Don't really see a reason for passing in the propertykey_type here, The type shouldn't change, so we can can just use `f.oracle.binding.settlementDataPropertyType`, no?

File: core/products/future_test.go [hunk 31]
```diff
@@ -36,47 +36,121 @@ func TestScalingOfSettlementData(t *testing.T) {
...
+	scaled, err := ft.future.ScaleSettlementDataToDecimalPlaces(
+		n, 5,
+	)
 	require.NoError(t, err)
 	require.Equal(t, num.NewUint(100000), scaled)
+
+	// Create test future with settlement data type decimal with decimals (that represents a decimal)
+	ft = testFuture(t, datapb.PropertyKey_TYPE_DECIMAL)
+
+	// settlement data is in 5 decimal places, asset in 3 decimal places => x10^-2
```
Comment (EVODelavega) [inline core/products/future_test.go:54]: comment doesn't apply anymore. Perhaps we should return an error if the settlement data has more decimals than the asset supports (ie settlement data in ETH can never be valid if it has more than 18 decimal places)

File: core/products/future_test.go (hunk 15 above)
Comment (EVODelavega) [reply core/products/future_test.go:106]: `10000.00001` being the settlement data for an asset with only 3 decimal places doesn't make sense. If the settlement data is a value in asset X, it can only be valid if it has no more than 3 decimals.

File: core/types/data_source.go [hunk 32]
```diff
@@ -2,13 +2,16 @@ package types
...
 	"encoding/hex"
+	"errors"
 	"fmt"
 
 	"code.vegaprotocol.io/vega/libs/crypto"
 	"code.vegaprotocol.io/vega/libs/proto"
 	vegapb "code.vegaprotocol.io/vega/protos/vega"
 )
 
+var ErrMultipleSameKeyNamesInFilterList = errors.New("multiple keys with same name found in filter list")
```
Comment (EVODelavega) [inline core/types/data_source.go:13]: This error is duplicated in oracle spec I believe, we should probably use the same error there

File: core/products/future.go (hunk 27 above)
Comment (AnExsomnis) [reply core/products/future.go:77]: This has been addresses, thanks

File: core/products/future.go (hunk 29 above)
Comment (AnExsomnis) [reply core/products/future.go]: Yes, this is fixed. Thanks

File: core/products/future_for_test.go (hunk 30 above)
Comment (AnExsomnis) [reply core/products/future_for_test.go:24]: Yes, sorry I missed that.
Corrected now.

File: core/types/data_source.go (hunk 32 above)
Comment (AnExsomnis) [reply core/types/data_source.go:13]: Yes, the problem here was that oracles package is already importing `types`, so here I will definitely not import oracles because I will create a cycle.
I can put the error only in this package and the oracle to import it.
But the error contains oracle logic domain.

File: core/products/future_test.go (hunk 31 above)
Comment (AnExsomnis) [reply core/products/future_test.go:54]: Okay, this is what I missed. I get it.
I'll correct the formula now and the tests.
Thanks!

File: libs/num/numeric.go (hunk 26 above)
Comment (EVODelavega) [reply libs/num/numeric.go]: In that comment I meant that the implementation returned the underlying pointer, so the clone of the original object would share a pointer to the same data of the original.

File: core/types/data_source.go (hunk 32 above)
Comment (EVODelavega) [reply core/types/data_source.go:13]: The oracles package is already importing the types package, and this error here is in the types package. So the error should only be defined here, and not in the oracles package. There's no issues with cyclical imports

Comment (AnExsomnis) [reply core/types/data_source.go:13]: Yes, I placed it in the types package :+1:

File: core/execution/market_snapshot.go (hunk 24 above)
Comment (AnExsomnis) [reply core/execution/market_snapshot.go]: Yes, I forgot to remove that.
Corrected now.

File: core/execution/market_snapshot.go (hunk 23 above)
Comment (AnExsomnis) [reply core/execution/market_snapshot.go]: Right, set it to be a pointer in both places. Thanks!

File: libs/num/numeric.go (hunk 26 above)
Comment (AnExsomnis) [reply libs/num/numeric.go]: Understood, changed that to have a copy of the value and return pointer to it -> https://github.com/vegaprotocol/vega/pull/6935/commits/45c9f3f7de3c29edb1d92a134ee7e1e34f567730#diff-01e84bbc9c49dbdd4b9e45ec724423ea40a7c808577eebb58ed4f69f25ac4755R110-R124
Thanks!

File: libs/num/numeric.go (hunk 25 above)
Comment (AnExsomnis) [reply libs/num/numeric.go]: Please see comment https://github.com/vegaprotocol/vega/pull/6935#discussion_r1050996252

File: core/products/future.go (hunk 28 above)
Comment (AnExsomnis) [reply core/products/future.go]: Yes, this is a mistake, sorry.
Corrected.

**Review by EVODelavega — APPROVED** [review]

**Review by vega-paul — APPROVED** [review]

**Review by candida-d — APPROVED** [review]
